lib/kubernetes-deploy/kubectl.rb

```diff
@@ -20,26 +21,35 @@ def initialize(namespace:, context:, logger:, log_failure_by_default:, default_t
       raise ArgumentError, "context is required" if context.blank?
     end
 
-    def run(*args, log_failure: nil, use_context: true, use_namespace: true, raise_if_not_found: false)
+    def run(*args, log_failure: nil, use_context: true, use_namespace: true, raise_if_not_found: false, attempts: 1)
       log_failure = @log_failure_by_default if log_failure.nil?
 
       args = args.unshift("kubectl")
       args.push("--namespace=#{@namespace}") if use_namespace
       args.push("--context=#{@context}")     if use_context
       args.push("--request-timeout=#{@default_timeout}") if @default_timeout
-
-      @logger.debug Shellwords.join(args)
-      out, err, st = Open3.capture3(*args)
-      @logger.debug(out.shellescape) unless output_is_sensitive?
-
-      unless st.success?
-        if log_failure
-          @logger.warn("The following command failed: #{Shellwords.join(args)}")
-          @logger.warn(err) unless output_is_sensitive?
-        end
-
-        if raise_if_not_found && err.match(NOT_FOUND_ERROR_TEXT)
-          raise ResourceNotFoundError, err
+      out, err, st = nil
+
+      (1..attempts).to_a.each do |attempt|
+        @logger.debug "Running command (attempt #{attempt}): #{args.join(' ')}"
+        out, err, st = Open3.capture3(*args)
+        @logger.debug("Kubectl out: " + out.gsub(/\s+/, ' ')) unless output_is_sensitive?
+
+        if st.success?
+          break
+        else
+          if log_failure
+            @logger.warn("The following command failed: #{Shellwords.join(args)}")
```

Comment:
I think we should add the something like `attempt: #{attempt}/#{attempts}` to this message.

Reply:
The attempt number is already recorded at the beginning of each attempt, so you see it whether the command ultimately succeeds or fails: `@logger.debug "Running command (attempt #{attempt}): #{args.join(' ')}"`

Reply:
It just seemed odd to me that the test in `test_run_with_multiple_attempts_retries_and_emits_failure_metrics` doesn't actually have the attempts in the output

Reply:
It does though:

![image](https://user-images.githubusercontent.com/4789493/49045961-987a1880-f1a0-11e8-9d31-8884e17f62d8.png)